gcc/rust/typecheck/rust-hir-type-check-path.cc

```diff
@@ -179,20 +183,72 @@ void
 TypeCheckExpr::visit (HIR::PathInExpression &expr)
 {
   NodeId resolved_node_id = UNKNOWN_NODEID;
-  size_t offset = -1;
-  TyTy::BaseType *tyseg = resolve_root_path (expr, &offset, &resolved_node_id);
-  if (tyseg->get_kind () == TyTy::TypeKind::ERROR)
-    return;
-
-  bool fully_resolved = offset == expr.get_segments ().size ();
-  if (fully_resolved)
+  if (expr.get_path_kind () == HIR::PathPattern::Kind::LangItem)
     {
-      infered = tyseg;
-      return;
+      auto lookup
+	= Analysis::Mappings::get ().get_lang_item_node (expr.get_lang_item ());
+      auto hir_id = mappings.lookup_node_to_hir (lookup);
+
+      // We can type resolve the path in expression easily as it is a lang
+      // item path, but we still need to setup the various generics and
+      // substitutions
+
+      // FIXME: We probably need to check *if* the type needs substitutions
+      // or not
+      if (LangItem::IsEnumVariant (expr.get_lang_item ()))
+	{
+	  std::pair<HIR::Enum *, HIR::EnumItem *> enum_item_lookup
+	    = mappings.lookup_hir_enumitem (*hir_id);
+	  bool enum_item_ok = enum_item_lookup.first != nullptr
+			      && enum_item_lookup.second != nullptr;
+	  rust_assert (enum_item_ok);
+
+	  HirId variant_id
+	    = enum_item_lookup.second->get_mappings ().get_hirid ();
+
+	  HIR::EnumItem *enum_item = enum_item_lookup.second;
+	  resolved_node_id = enum_item->get_mappings ().get_nodeid ();
+
+	  // insert the id of the variant we are resolved to
+	  context->insert_variant_definition (expr.get_mappings ().get_hirid (),
+					      variant_id);
+
+	  query_type (variant_id, &infered);
+	  infered = SubstMapper::InferSubst (infered, expr.get_locus ());
+	}
```

Comment:
This is super dodgy, @philberty please check it good lol

Reply:
If I could ask, how could you end up with a lang-item path that DOESN'T need substitution inferencing? I'm curious because I think the only way you could generate an option type that doesn't need inferencing would be to generate a lang-item path like Option::<i32>::Some(3), but that wouldn't be possible since lang-item paths wouldn't have PathExprSegments which can be generic arguments

Reply:
I think it's probably better to keep it I would imagine, since if it ever becomes an issue, the comment clarifies that ALWAYS doing substitution inference is not a design choice

Reply:
there might be a change needed for the generics but would need to try it out a bit myself to try our some tests on it